File: example/petstore/src/Petstore.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}

module Petstore where

import Control.Monad
import Control.Monad.IO.Class
import Data.Aeson
import qualified Data.ByteString.Builder as BSB
import qualified Data.ByteString.Lazy as LBS
import Data.IORef
import Data.List
import GHC.Generics (Generic)
import Network.Wai
import Network.Wai.Handler.Warp
import Servant

------------------------------------------------------------------------

type PetstoreAPI =
  "pet" :> Capture "petId" Int :> Get '[JSON] Pet
    :<|> "pet" :> "badJson" :> Capture "petId" Int :> Get '[JSON] Pet
    :<|> "pet" :> ReqBody '[JSON] Pet :> Post '[JSON] ()
    :<|> "health" :> Get '[JSON] ()
    :<|> "_reset" :> Delete '[JSON] ()

data Pet = Pet
  { petId :: Int
  , petName :: String
  }
  deriving (Eq, Show, Generic)

instance ToJSON Pet
instance FromJSON Pet

server :: IORef [Pet] -> Server PetstoreAPI
server store = getPet :<|> badGetPet :<|> addPet :<|> health :<|> reset
  where
    getPet :: Int -> Handler Pet
    getPet pid = do
      pets <- liftIO (readIORef store)
      case find (\pet -> petId pet == pid) pets of
        Nothing -> throwError err404
        Just pet -> return pet

    -- XXX: We'll make it do bad stuff in the middleware.
    badGetPet :: Int -> Handler Pet
    badGetPet = getPet

    addPet :: Pet -> Handler ()
    addPet pet = do
      pets <- liftIO (readIORef store)
      when (pet `elem` pets) $
        throwError err409 {errBody = "Pet already exists"}
      liftIO (writeIORef store (pet : pets))

    health :: Handler ()
    health = return ()

    reset :: Handler ()
    reset = liftIO (writeIORef store [])

petstoreAPI :: Proxy PetstoreAPI
petstoreAPI = Proxy

app :: IORef [Pet] -> Application
app store = faultInject (serve petstoreAPI (server store))

faultInject :: Middleware
faultInject baseApp req respond0
  | "badJson" `elem` pathInfo req =
      baseApp req $ \response -> do
        let status = responseStatus response
        let headers = responseHeaders response
        body <- responseBody response
        respond0 $ responseLBS status headers (LBS.drop 1 body)
  | otherwise = baseApp req respond0

responseBody :: Response -> IO LBS.LazyByteString
responseBody resp =
  let (_status, _headers, body) = responseToStream resp
  in  body $ \f -> do
        content <- newIORef mempty
        f (\chunk -> modifyIORef' content (<> chunk)) (return ())
        BSB.toLazyByteString <$> readIORef content

libMain :: Port -> IO ()
libMain port = do
  store <- newIORef [Pet 1 "apa"]
  run port (app store)

File: example/petstore-python-flask/petstore.py
from flask import Flask
from flask import jsonify
from flask import request

app = Flask(__name__)

# NOTE: Not thread-safe and won't work if clients connect concurrently.
petstore = {}

@app.route("/pet/<pet_id>", methods=["GET"])
def get_pet(pet_id):
    if pet_id in petstore:
        pet_name = petstore[pet_id]
        return jsonify(petId=pet_id, petName=pet_name)
    else:
        return "Pet not found", 404

@app.route("/pet", methods=["POST"])
def add_pet():
    pet_id = request.json.get("petId", None)
    pet_name = request.json.get("petName", None)
    if pet_id in petstore:
        return jsonify(error="Pet already exists"), 409
    petstore[pet_id] = pet_name
    return "[]", 201

@app.route("/health", methods=["GET"])
def health():
    return "[]", 200

@app.route("/_reset", methods=["DELETE"])
def _reset():
    petstore.clear()
    return "[]", 200

if __name__ == "__main__":
    app.run()
